Testing.Rmd

```diff
@@ -442,159 +769,1106 @@ It's hard to give good general advice about writing tests, but you might find th
-  base <- as.POSIXct("2009-08-03 12:01:59.23", tz = "UTC")
-  floor_base <- function(unit) floor_date(base, unit)
-  as_time <- function(x) as.POSIXct(x, tz = "UTC")
-
-  expect_equal(floor_base("second"), as_time("2009-08-03 12:01:59"))
-  expect_equal(floor_base("minute"), as_time("2009-08-03 12:01:00"))
-  expect_equal(floor_base("hour"),   as_time("2009-08-03 12:00:00"))
-  expect_equal(floor_base("day"),    as_time("2009-08-03 00:00:00"))
-  expect_equal(floor_base("week"),   as_time("2009-08-02 00:00:00"))
-  expect_equal(floor_base("month"),  as_time("2009-08-01 00:00:00"))
-  expect_equal(floor_base("year"),   as_time("2009-01-01 00:00:00"))
+test_that("multiplication works", {
+  useful_thing <- 3
+  expect_equal(2 * useful_thing, 6)
+})
+
+test_that("subtraction works", {
+  useful_thing <- 3
+  expect_equal(5 - useful_thing, 2)
 })
 ```
 
-We could go a step further and create a custom expectation function:
+In real life, `useful_thing` is usually a more complicated object that somehow feels burdensome to instantiate.
+Notice how `useful_thing <- 3` appears in more than once place.
+Conventional wisdom says we should DRY this code out.
+It's tempting to just move `useful_thing`'s definition outside of the tests:
 
 ```{r}
-base <- as.POSIXct("2009-08-03 12:01:59.23", tz = "UTC")
+useful_thing <- 3
+
+test_that("multiplication works", {
+  expect_equal(2 * useful_thing, 6)
+})
+
+test_that("subtraction works", {
+  expect_equal(5 - useful_thing, 2)
+})
+```
+
+This does work because when `useful_thing` is not found in the test-specific environment, the search continues in the parent environment, where `useful_thing` will often be found.
+When testthat executes an entire test file, `useful_thing` will be defined and made available for subsequent tests in that file.
+However, during interactive development, there is no gesture to systematically identify and execute such code.
+This is particularly frustrating when the failing test starts on line 573 and all the implicit setup code is at line 1.
+Or, even worse, the setup is sprinkled around the file between tests, at random locations that made sense at the time you wrote it.
+
+> In its purest form, automating testing consists of three activities: writing tests, running tests, and **reacting to test failures**.
+> 
+> From the book Software Engineering at Google, [Chapter 11](https://abseil.io/resources/swe-book/html/ch11.html)
+
+Top-level code, outside of any test, is in a sort of No Man's Land:
+it's fine when your tests are all passing, but when something goes wrong and a test fails, now you have to endure extra self-inflicted pain when you least need it.
+This is the sort of practice that feels convenient when *writing tests*, but proves to be extremely inconvenient when you are *failing tests*.
+Below we recommend various robust and structured solutions to this problem.
+
+But first, seriously consider inlining the creation of a `useful_thing` whenever you need it.
```

Comment:
why not mention here the idea to make its creation a function defined centrally, if creating it takes too much space?